### PR TITLE
Respect list of source directories

### LINKED ISSUE #1264: [v0.16] mutating undesired code
| Question    | Answer
| ------------| ---------------
| Infection version | 0.16.3
| Test Framework version | PHPUnit 8.5.4
| PHP version | 7.4.5 
| Platform    | e.g. Ubuntu 16.04
| Github Repo | -

In order to run infection with 100% msi, we break our business logic into a "functional core" which can be thoroughly unit tested and kill every mutant. Then we run infection only over the functional core by using the `"--testsuite=unit"` test framework option and by using `source.directories` in `infection.json.dist`. With version 0.15.* this works exactly as expected, with mutants only being applied to our functional core:

* 178 files mutated
* 940 mutants generated

With version 0.16.3, infection is mutating our entire codebase:

* 1578 files mutated
* 10211 mutants generated
 

<details>
 <summary>phpunit.xml</summary>

 ```xml
<?xml version="1.0" encoding="UTF-8"?>

<!-- https://phpunit.de/manual/current/en/appendixes.configuration.html -->
<phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:noNamespaceSchemaLocation="http://schema.phpunit.de/8.3/phpunit.xsd"
         backupGlobals="false"
         colors="true"
         bootstrap="config/bootstrap.php"
>
    <php>
        <ini name="error_reporting" value="-1" />
        <server name="APP_ENV" value="test" force="true" />
        <server name="SYMFONY_PHPUNIT_VERSION" value="8.3.5" />
        <server name="SHELL_VERBOSITY" value="-1" />
    </php>

    <testsuites>
        <testsuite name="All">
            <directory>tests/</directory>
        </testsuite>

        <!-- Different test types you might run separately -->
        <testsuite name="Unit">
            <directory>tests/Unit/</directory>
        </testsuite>
        <testsuite name="Integration">
            <directory>tests/Integration/</directory>
        </testsuite>
        <testsuite name="Message">
            <directory>tests/Message/</directory>
        </testsuite>
        <testsuite name="Contract">
            <directory>tests/Contract/</directory>
        </testsuite>
        <testsuite name="Model">
            <directory>tests/Model</directory>
        </testsuite>
        <testsuite name="Task">
            <directory>tests/Task</directory>
        </testsuite>

        <testsuite name="Functional">
            <directory>tests/Functional/</directory>
        </testsuite>

        <testsuite name="FunctionalAdmin">
            <directory>tests/Functional/Admin</directory>
        </testsuite>
        <testsuite name="FunctionalExport">
            <directory>tests/Functional/Export</directory>
        </testsuite>
        <testsuite name="FunctionalPlaybook">
            <directory>tests/Functional/Playbook</directory>
        </testsuite>
        <testsuite name="FunctionalUser">
            <directory>tests/Functional/User</directory>
        </testsuite>
        <testsuite name="FunctionalNegotiate">
            <directory>tests/Functional/Negotiate</directory>
        </testsuite>
        <testsuite name="FunctionalActivityStream">
            <directory>tests/Functional/ActivityStream</directory>
        </testsuite>
    </testsuites>

    <filter>
        <whitelist>
            <directory>./src/</directory>
        </whitelist>
    </filter>

    <listeners>
        <listener class="Hyra\Tests\Resources\TestListener\TestTidierListener" />
    </listeners>
</phpunit>

 ```
</details>

I tried stepping through with xdebug to track down the issue but got lost as so much has changed from 0.15 to 0.16. Happy to keep debugging if someone can help point me in the right direction?

### PR DESCRIPTION
**Pull request by sanmai**
This PR:

- [x] Fixes #1264
- [x] E2E tests are passing
- [x] Covered by tests
- [x] CI is passing

### CODE REVIEW

**Comment by maks-rafalko** [comment]
Thanks @sanmai for looking into it.

Just checked how it works, the issue is resolved but partially as I can see.

* `bin/infection` - works correctly now 👍 
* `bin/infection --only-covered` works incorrectly if I add the following diff 

```diff
diff --git a/tests/e2e/Source_Directories_Config/tests/SourceClassTest.php b/tests/e2e/Source_Directories_Config/tests/SourceClassTest.php
index 2a2bf2fc..df1737e2 100644
--- a/tests/e2e/Source_Directories_Config/tests/SourceClassTest.php
+++ b/tests/e2e/Source_Directories_Config/tests/SourceClassTest.php
@@ -4,6 +4,7 @@ namespace Source_Directories_Config\Test;

 use Source_Directories_Config\ToMutate\SourceClass;
 use PHPUnit\Framework\TestCase;
+use Source_Directories_Config\ToNotMutate\ToNotMutateSourceClass;

 class SourceClassTest extends TestCase
 {
@@ -12,4 +13,10 @@ class SourceClassTest extends TestCase
         $sourceClass = new SourceClass();
         $this->assertSame('hello', $sourceClass->hello());
     }
+
+public function test_to_not_mutate()
+{
+    $sourceClass = new ToNotMutateSourceClass();
+    $this->assertSame('to not mutate', $sourceClass->toNotMutate());
+}
 }
```

In other words, if the code from `src/ToNotMutate` is covered by tests, and with only `"src/ToMutate"` folder in `infection.json`, then when Infection is executed with `--only-covered`, source directory config is not respected.

I would say we need to filter out files a bit earlier, somewhere here https://github.com/infection/infection/pull/1269/files#diff-2a3ea38300a250b7fd2256922fc32392R103

**Comment by sanmai** [comment]
You're 100% correct that traces being added in a wrong order. Guess I should have been more clear on why I left it as draft. I'll see about finishing this a bit later, and only then I'll tag it as ready for review.

**Comment by sanmai** [comment]
@maks-rafalko The overall architecture seems to be ready. E2E tests are passing. If you can see this refactoring I did makes sense, I will proceed with finishing with tests and CI.

File: src/TestFramework/Coverage/BufferedSourceFileFilter.php [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+                }
+
+                return false;
+            });
+    }
+
+    /**
+     * @return SplFileInfo[]
+     */
+    public function getUnseenFiles(): array
```
Comment (maks-rafalko) [inline src/TestFramework/Coverage/BufferedSourceFileFilter.php:111]: only after debugging with xdebug I understood what this is used for. What do you think about renaming it to for example `unseenInCoverageReportFiles`.

As far as I understood, this method returns files that **are** in `source.directories` from `infection.json.dist` but **not** in coverage report (phpunit's `filter.whitelist.directory`)

Comment (sanmai) [reply src/TestFramework/Coverage/BufferedSourceFileFilter.php:111]: That's exactly what it is and I will rename it to match.

**Comment by sanmai** [comment]
The build is green. 

Shall we merge now and get to these extra tests in the next PR? (With an open issue to track efforts.)

**Comment by maks-rafalko** [comment]
> Shall we merge now and get to these extra tests in the next PR? (With an open issue to track efforts.)

sure, let's do it. Could you please try targeting to 0.16 so we can release a bugfix patch version? 

If it's hard to do - np, let's just merge to `master`

**Comment by sanmai** [comment]
Well, it doesn't seem to be straightforward. I think it'll be easier to cherry-pick this commit later.

**Comment by maks-rafalko** [comment]
Thank you, @sanmai

**Pull request merged**